README.md
CLI to create video mosaics.

### Installation

```
make install
```

If you're on Linux a mpv user script can be create with:

```
make mpv
```

### Usage

```
genthumbs [videos]
```

### Dependencies
 - `ffmpeg`
 - `ffprobe`
 - `zx`
 - `xxh128sum (optional)`
 
 ### Examples

 ![BigBuckBunny](/assets/BigBuckBunny.jpg)
 ![owarimonogatari](/assets/owarimonogatari.jpg)
 ![bonobo](/assets/bonobo.jpg)


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

## [1.2.0](https://github.com/gqgs/genthumbs/compare/v1.1.0...v1.2.0) (2021-06-06)


### Features

* add --nohash option ([fe5ee3e](https://github.com/gqgs/genthumbs/commit/fe5ee3ed5e3caa747a355410b9f65981bef97470))
* add --open option ([c29a8e5](https://github.com/gqgs/genthumbs/commit/c29a8e5c72d0561958bb8d5b46611cde7dc7bafd))
* add rule to create mpv user script ([1322d15](https://github.com/gqgs/genthumbs/commit/1322d159655e7a63b134e2abfa5a91024d33a40c))

## [1.1.0](https://github.com/gqgs/genthumbs/compare/v1.0.0...v1.1.0) (2021-05-31)


### Features

* add draw timestamp option ([f099c02](https://github.com/gqgs/genthumbs/commit/f099c02f4800cc1a73f8fa17656e5bee8aa38c65))
* port it to zx ([b7bdbeb](https://github.com/gqgs/genthumbs/commit/b7bdbebccd1afab7405b0e030bd415ae41364a5c))

## 1.0.0 (2020-12-17)


### Features

* add help command ([fbe527e](https://github.com/gqgs/genthumbs/commit/fbe527e8676ac68bdcac5e73fbd2b26f2ee0652e))
* use thumbnail filter to help select frame ([0903159](https://github.com/gqgs/genthumbs/commit/09031591ef35e721967174e63da2551ec8418079))


### Bug Fixes

* avoid internal consumption of stdin ([2f9dde7](https://github.com/gqgs/genthumbs/commit/2f9dde7181b8b77bfb7fba27885c737db1592fce))
* avoid processing invalid files ([6cebb5b](https://github.com/gqgs/genthumbs/commit/6cebb5bb847511e821e01b4e9806ba484ad393d8))
* filename handling ([a6f3e0b](https://github.com/gqgs/genthumbs/commit/a6f3e0b1e62490f515773840fac557294bb45833))
* improve output quality ([c814c24](https://github.com/gqgs/genthumbs/commit/c814c2410e1331281ba6b67208cdf989c94a261b))
* insufficient frames for small files ([398457a](https://github.com/gqgs/genthumbs/commit/398457a92f189cdfea2404ddf008185837fdb285))
* preserve natural order while creating tile ([0af1e2d](https://github.com/gqgs/genthumbs/commit/0af1e2d6cb31d2bc96e8e87500f050180ee7eb80))
* remove unused variable ([68c036b](https://github.com/gqgs/genthumbs/commit/68c036b99d6a348c2d81a81129a9f3e4a32ef72e))


thumbs.sh
#!/usr/bin/bash

usage() {
    echo "Usage: genthumbs [videos]"
    echo "-t: draw timestamps"
    echo
    exit 1
}

while getopts "t" opt; do
    case "${opt}" in
    t)
        timetext=",drawtext=text='\$text':fontsize=15:y=10:x=10:box=1:boxcolor=white@0.15"
        shift
        ;;
    *)
        usage
        ;;
    esac
done

if [ $# -eq "0" ]; then
    usage
fi

function parsetime() {
    date="@$1"
    echo $(date -u --date=$date +"%T")
}

for file in "$@"; do
    echo "processing: '$file'"
    duration=$(ffprobe "$file" -show_entries format=duration -v quiet -of csv="p=0")
    duration=${duration/.*}

    if [[ $duration == "N/A" ]]; then
        echo "invalid duration '$duration'"
        continue
    fi

    timestamp=$(parsetime $duration)

    size=$(ls -sh "$file")
    size=${size/\ *}

    digest=$(xxh128sum "$file")
    digest=${digest/\ *}

    text="$size / $timestamp / $digest"
    text=${text//":"/"\:"}

    tmpdir=$(mktemp -d)

    start=$(bc <<< "scale=2; $duration/20")
    pos=$start
    while [ "$(bc <<< "$pos < $duration")" == "1" ]
    do
        pos=$(bc <<< "scale=2; $pos + $start")
        n=$(bc <<< "scale=0; 20*$pos/$duration")
        possuffix=`printf "%010.f" $n` # padding needed to keep natural sorting order

        timestamp=$(parsetime $pos)
        timestr=${timetext/"\$text"/"$timestamp"}
        timestr=${timestr//":"/"\:"}

        ffmpeg -nostdin -loglevel error -ss "$pos"s -i "$file" -vf thumbnail=10,scale=320:-1$timestr -vsync 0 -q:v 1 -frames:v 1 -y "$tmpdir/${file/.*/_$possuffix.jpg}" &
    done
    wait
    ffmpeg -nostdin -loglevel error -pattern_type glob -i "$tmpdir/*.jpg" -vf tile=4x4:color=white:padding=5:margin=50,drawtext="text='$text':fontsize=30:y=10:x=50" -vsync 0 -frames:v 1 -q:v 1 -y "${file/.*/.jpg}"
    rm -r "$tmpdir"
done

thumbs.mjs
#!/usr/bin/env zx

const usage = () => {
    console.log("Usage: genthumbs [videos]")
    console.log("--open open the thumbnail after creation")
    process.exit(2)
}

const args = process.argv.splice(3)
const help = args.filter(arg => arg === "--help").length > 0
const open = args.filter(arg => arg === "--open").length > 0
const nohash = args.filter(arg => arg === "--nohash").length > 0
const files = args.filter(arg => !arg.startsWith("-"))

if (files.length == 0 || help) {
    usage()
}

const gridText = async (file, size, timestamp) => {
    if (nohash) return `${size} / ${timestamp}`
    const digest = await $`xxh128sum ${file} 2>/dev/null | awk '{print $1}'`
    return `${size} / ${timestamp} / ${digest}`
}

files.forEach(async file => {
    const duration = await $`ffprobe ${file} -show_entries format=duration -v quiet -of csv="p=0"`
    const timestamp = await $`date -u --date=@${duration} +"%T"`
    const size = await $`ls -sh ${file} | awk '{print $1}'`

    const text = (await gridText(file, size, timestamp)).replaceAll(":", "\\:").replaceAll("\n", "")
    const tmpdir = await $`mktemp -d`
    const step = parseFloat(duration)/20
    const jobs = []

    const basedir = await $`dirname ${file}`
    const filename = await $`basename ${file}`
    const output = `${basedir}/${filename}.jpg`.replaceAll("\n", "")

    for (let pos = step, i = 0; pos < duration; pos += step, i++) {
        const index = i.toString().padStart(4, 0)
        jobs.push($`ffmpeg -nostdin -loglevel error -ss "${pos}"s -i ${file} -vsync 0 -q:v 1 -frames:v 1 -y "${tmpdir}/${index}.jpg"`)
    }

    await Promise.all(jobs)
    await $`ffmpeg -nostdin -loglevel error -pattern_type glob -i "${tmpdir}/*.jpg" \
            -vf tile=4x4:color=white:padding=5:margin=50,drawtext="text=${text}:fontsize=30:y=10:x=50" \
            -vsync 0 -frames:v 1 -q:v 1 -y ${output}`

    await $`rm -rf ${tmpdir}`

    if (open) $`xdg-open ${output}`
})